controller/controlplane/manager_options.go

```diff
@@ -93,3 +96,105 @@ func WithMetricsServerOff() managercfg.Opt {
 		c.MetricsAddr = "0" // 0 disables metrics server
 	}
 }
+
+// WithFeatureGates sets the feature gates for the manager.
+func WithFeatureGates(logger logr.Logger, featureGates []gwtypes.ControlPlaneFeatureGate) managercfg.Opt {
+	return func(c *managercfg.Config) {
+		fgs := managercfg.FeatureGates{}
+		defaults := managercfg.GetFeatureGatesDefaults()
+		for _, feature := range featureGates {
+			if _, ok := defaults[feature.Name]; !ok {
+				log.Error(logger, errors.New("unknown feature gate"), "unknown feature gate",
+					"feature", feature.Name, "state", feature.State,
+				)
+				continue
+			}
+
+			// This should never happen as it should be enforced at the CRD level
+			// but we handle it gracefully here and log an error.
+			if _, ok := fgs[feature.Name]; ok {
+				log.Error(logger, errors.New("feature gate already set"), "feature gate already set",
+					"feature", feature.Name, "state", feature.State,
+				)
+				continue
+			}
+			fgs[feature.Name] = (feature.State == gwtypes.FeatureGateStateEnabled)
+		}
+
+		for k, v := range defaults {
+			// Ensure that we don't override the defaults with empty values
+			if _, ok := fgs[k]; !ok {
+				fgs[k] = v
+			}
+		}
+		c.FeatureGates = fgs
+	}
+}
+
+// WithControllers sets the controllers for the manager.
+func WithControllers(logger logr.Logger, controllers []gwtypes.ControlPlaneController) managercfg.Opt {
+	logDeprecated := func(logger logr.Logger, enabled bool, controllerName string) {
+		if enabled {
+			log.Info(logger, "chosen controller is deprecated", "controller", controllerName)
+		}
+	}
+	setOpt := func(b *bool, state gwtypes.ControllerState) {
+		if b == nil {
+			return
+		}
+		*b = (state == gwtypes.ControlPlaneControllerStateEnabled)
+	}
+	return func(c *managercfg.Config) {
+		for _, controller := range controllers {
+			switch controller.Name {
+			case "INGRESS_NETWORKINGV1":
+				setOpt(&c.IngressNetV1Enabled, controller.State)
+			case "INGRESS_CLASS_NETWORKINGV1":
+				setOpt(&c.IngressClassNetV1Enabled, controller.State)
+			case "INGRESS_CLASS_PARAMETERS":
+				setOpt(&c.IngressClassParametersEnabled, controller.State)
+
```

Comment:
Are the empty lines added by purpose to group the configuration items? If so, I suggest to add comments to tell how did we group the items.

Reply:
👍 d9f98d710dc42ad6c2829a71ec65ac2bffbe2835